In [19]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [20]:
# Import departments.csv dataset

path = r'C:\Users\sanch\Documents\Instacart Basket Analysis'
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col = False)

In [21]:
# View the first five rows of df_dep.csv dataframe

df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

#### **Transpose**

In [22]:
# Transpose df_dep.csv and rename it

df_dep_t = df_dep.T

#### **Reset the index**

In [23]:
# Change the index for df_dep_t.csv dataset

df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [24]:
# Extract the new header for df_dep.csv

new_header = df_dep_t.iloc[0]

In [25]:
# Extract the first row and beyond from df_dep_t.csv 

df_dep_t_new = df_dep_t[1:]

In [26]:
# Assign the new header

df_dep_t_new.columns = new_header

In [27]:
# View the df_dep_t_new.csv dataset

df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [28]:
# Create a data dictionary for df_dep_t_new dataset

data_dict_dep = df_dep_t_new.to_dict('index')

In [29]:
# View the data dictionary

data_dict_dep

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [30]:
# Import products.csv dataset

df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [31]:
# View the first 5 rows of df_prods dataframe

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [32]:
# Create a subset for df_prods dataframe that only contains data from the snacks department

df_snacks = df_prods[df_prods['department_id'] == 19]

In [33]:
# View the first five rows of the subsetted df_snacks dataframe

df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

### Tasks

In [34]:
# Import orders.csv dataset

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

#### **Change data types**

In [35]:
# Change the data type of order_id to 'string' for df_ords

df_ords['order_id'] = df_ords['order_id'].astype('str')

#### **Change column name**

In [36]:
# Change the column name from 'order_dow' to 'order_day_of_week'

df_ords.rename(columns = {'order_dow' : 'order_day_of_week'}, inplace = True)

#### **Find the busiest hour placing orders**

In [37]:
# Count number of order for hour of day

df_ords['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: count, dtype: int64

**Summary:** The busiest hour is for placing orders is 10:00.

#### Determine the meaning behind 'department_id' = 4 column within the df_prods dataframe using a data dictionary.

In [38]:
print(data_dict_dep.get('4'))

{'department': 'produce'}


The "department_id"  = 4 column is about produce.

#### Create a subset containing only the breakfast item sales. 

In [39]:
# Create a data dictionary for df_prods.csv dataset

data_dict_prods = df_prods.to_dict('index')

In [40]:
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [41]:
df_breakfast.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

#### Find all observations that include items from the following departments: alcohol, deli, beverages, and meat/seafood.

In [42]:
df_mixed = df_prods.loc[df_prods['department_id'].isin([5,7,12,20])]

In [43]:
df_mixed.head()

product_id                                    product_name  aisle_id  \
2            3            Robust Golden Unsweetened Oolong Tea        94   
6            7                  Pure Coconut Water With Orange        98   
9           10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                               Peach Mango Juice        31   
16          17                               Rendered Duck Fat        35   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1

#### How many rows does the last dataframe you created have?

In [44]:
df_mixed.shape

(7650, 5)

#### **Summary:** The df_mixed dataframe contains 7650 rows.

#### Extract all the information about this user "user_id = 1"

In [45]:
df_user_1 = df_ords[df_ords['user_id']==1]

In [46]:
df_user_1.head()

order_id  user_id eval_set  order_number  order_day_of_week  \
0  2539329        1    prior             1                  2   
1  2398795        1    prior             2                  3   
2   473747        1    prior             3                  3   
3  2254736        1    prior             4                  4   
4   431534        1    prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

#### Provide some details about this user’s behavior. 

In [47]:
df_user_1.describe()

user_id  order_number  order_day_of_week  order_hour_of_day  \
count     11.0     11.000000          11.000000          11.000000   
mean       1.0      6.000000           2.636364          10.090909   
std        0.0      3.316625           1.286291           3.477198   
min        1.0      1.000000           1.000000           7.000000   
25%        1.0      3.500000           1.500000           7.500000   
50%        1.0      6.000000           3.000000           8.000000   
75%        1.0      8.500000           4.000000          13.000000   
max        1.0     11.000000           4.000000          16.000000   

       days_since_prior_order  
count               10.000000  
mean                19.000000  
std                  9.030811  
min                  0.000000  
25%                 14.250000  
50%                 19.500000  
75%                 26.250000  
max                 30.000000

In [48]:
# Export df_ords dataframe to 'orders_wrangled.csv'

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [49]:
# Export df_dep_t_new dataframe to 'departments_wrangled.csv'

df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))